ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/ActivePeers.hs

```diff
@@ -207,31 +209,52 @@ jobPromoteWarmPeer PeerSelectionActions{peerStateActions = PeerStateActions {act
                    peeraddr peerconn =
     Job job handler "promoteWarmPeer"
   where
+    -- TODO: The delay should be randomly picked from an interval.
+    reconnectDelay :: DiffTime
+    reconnectDelay = 10
+
     handler :: SomeException -> Completion m peeraddr peerconn
     handler e =
-      --TODO: decide what happens if promotion fails, do we stay warm or go to
-      -- cold? Will this be reported asynchronously via the state monitoring?
+      -- When promotion fails we set the peer as cold.
       Completion $ \st@PeerSelectionState {
                                activePeers,
+                               establishedPeers,
+                               knownPeers,
                                targets = PeerSelectionTargets {
                                            targetNumberOfActivePeers
                                          }
                              }
-                    _now -> Decision {
+                    now ->
+        let establishedPeers' = EstablishedPeers.delete peeraddr
+                                  establishedPeers
+            knownPeers'       = if peeraddr `KnownPeers.member` knownPeers
+                                   then KnownPeers.setConnectTime
+                                          (Set.singleton peeraddr)
+                                          (reconnectDelay `addTime` now)
+                                        $ snd $ KnownPeers.incrementFailCount
+                                          peeraddr
+                                          knownPeers
+                                   else
+                                     -- Apparently the governor can remove
+                                     -- the peer we failed to promote from the
+                                     -- set of known peers before we can process
+                                     -- the failure.
+                                     knownPeers in
```

Comment:
We should make sure this is covered in a simulation test; either leave a todo or create an issue for that (or both :)).

Reply:
The code snippet is to cover a problem that the simulation test cases uncovered.